File: 0901-online-stock-span/0901-online-stock-span.cpp
class StockSpanner {
public:
    StockSpanner() {
        
    }
    
    int next(int price) {
        int cnt = 0;
        while (!st.empty() && price >= st.top()) {
            cnt += mp[st.top()];
            st.pop();
        }
        st.push(price);
        mp[price] = (cnt + 1);
        
        return mp[price];
    }
    
    private:
        map<int, int> mp;
        stack<int> st;
    };

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner* obj = new StockSpanner();
 * int param_1 = obj->next(price);
 */

File: 0901-online-stock-span/0901-online-stock-span.java
class StockSpanner {
    Stack<int[]> stack = new Stack<>();
    
    public StockSpanner() {
        
    }
    
    public int next(int price) {
        int ans = 0;
        while(!stack.isEmpty() && stack.peek()[0] <= price) {
            ans += stack.pop()[1];
        }
        stack.push(new int[] {price, ans + 1});
        return ans + 1;
    }
}

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner obj = new StockSpanner();
 * int param_1 = obj.next(price);
 */